Incorrect super() inference for classmethods calling classmethods that return instances (constructors)

### Steps to reproduce

Run the following:

```python
from astroid import extract_node
n = extract_node('''
class A:

    @classmethod
    def new(cls):
        return cls()

class B(A):
    @classmethod
    def new(cls):
        inst = super().new()
        return inst

B.new()
''')
print(next(n.infer()))
```

### Current Behavior

Outputs `Instance of .A`

### Expected behavior

In Python, `B.new()` is an instance of B. However, the above prints `Instance of .A`, so this should print `Instance of .B`. Digging in, this seems to be due to the inferred type of the return of the call to `super().new()` from `B.new()` being an instance of A, rather than an instance of B.

This leads to bugs like pycqa/pylint#981, where the returns from overridden constructor methods are inferred to be the wrong type.


### Version Tested

1.5.3


### What I have done

I have managed to get the above code snippet to output `Instance of .B` by updating the following section of `objects.py`: https://github.com/PyCQA/astroid/blob/7bfb5e83b162d2126c9929b6d863724176269b6d/astroid/objects.py#L154-L167

I have updated it to read as follows, with the reasoning that B's classmethod is really calling A's new() method with its class being passed to `super()`, not its instance:

```python
                # We can obtain different descriptors from a super depending
                # on what we are accessing and where the super call is.
                if self._scope.type == 'classmethod' and inferred.type == 'classmethod':
                    yield bases.BoundMethod(inferred, self.type)
                elif inferred.type == 'classmethod':
                    yield bases.BoundMethod(inferred, cls)
                elif self._scope.type == 'classmethod' and inferred.type == 'method':
                    yield inferred
                elif self._class_based or inferred.type == 'staticmethod':
                    yield inferred
                elif bases._is_property(inferred):
                    # TODO: support other descriptors as well.
                    for value in inferred.infer_call_result(self, context):
                        yield value
                else:
                    yield bases.BoundMethod(inferred, cls)
```

This breaks one test in the test suite:

```
======================================================================
FAIL: test_super_bound_model (astroid.tests.unittest_objects.SuperTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/Nautilus/Documents/Programming/astroid/astroid/tests/unittest_objects.py", line 285, in test_super_bound_model
    self.assertEqual(fourth.bound.name, 'First')
AssertionError: 'Super_Type_Type' != 'First'
- Super_Type_Type
+ First
```

Currently, I'm not certain if what I'm doing is correct. I intend to investigate further, to fix the broken test, and to add a test that illustrates the current pathological behavior. I don't know if I will have time this weekend (and I also don't really know what I'm doing!), so assistance is definitely welcome.

--- a/ChangeLog
+++ b/ChangeLog
@@ -36,6 +36,12 @@ Release date: TBA
 * Add simple command line interface for astroid to output generated AST.
   Use with ``python -m astroid``.
 
+* Fix incorrect type inference for ``super().method()`` calls that return ``Self``.
+  Previously, astroid would infer the parent class type instead of the child class type,
+  causing pylint E1101 false positives in method chaining scenarios.
+
+  Closes #457
+
 
 
 What's New in astroid 4.0.4?

--- a/astroid/bases.py
+++ b/astroid/bases.py
@@ -541,9 +541,13 @@ class BoundMethod(UnboundMethod):
         self,
         proxy: nodes.FunctionDef | nodes.Lambda | UnboundMethod,
         bound: SuccessfulInferenceResult,
+        original_caller: SuccessfulInferenceResult | None = None,
     ) -> None:
         super().__init__(proxy)
         self.bound = bound
+        # For super() calls: the actual instance/class that called super()
+        # Used to correctly infer return type of methods returning Self
+        self.original_caller = original_caller
 
     def implicit_parameters(self) -> Literal[0, 1]:
         if self.name == "__new__":
@@ -674,7 +678,9 @@ class BoundMethod(UnboundMethod):
         caller: SuccessfulInferenceResult | None,
         context: InferenceContext | None = None,
     ) -> Iterator[InferenceResult]:
-        context = bind_context_to_node(context, self.bound)
+        # For super() calls, use original_caller to correctly infer Self return type
+        bound_node = self.original_caller if self.original_caller else self.bound
+        context = bind_context_to_node(context, bound_node)
         if (
             isinstance(self.bound, nodes.ClassDef)
             and self.bound.name == "type"

--- a/astroid/objects.py
+++ b/astroid/objects.py
@@ -194,7 +194,8 @@ class Super(node_classes.NodeNG):
                 # We can obtain different descriptors from a super depending
                 # on what we are accessing and where the super call is.
                 if inferred.type == "classmethod":
-                    yield bases.BoundMethod(inferred, cls)
+                    # Pass original caller for classmethod too
+                    yield bases.BoundMethod(inferred, cls, original_caller=self.type)
                 elif self._scope.type == "classmethod" and inferred.type == "method":
                     yield inferred
                 elif self._class_based or inferred.type == "staticmethod":
@@ -214,7 +215,9 @@ class Super(node_classes.NodeNG):
                     except InferenceError:
                         yield util.Uninferable
                 else:
-                    yield bases.BoundMethod(inferred, cls)
+                    # Pass original caller (self.type) so infer_call_result can
+                    # correctly resolve Self return types to the actual caller type
+                    yield bases.BoundMethod(inferred, cls, original_caller=self.type)
 
         # Only if we haven't found any explicit overwrites for the
         # attribute we look it up in the special attributes

--- a/tests/test_objects.py
+++ b/tests/test_objects.py
@@ -504,6 +504,60 @@ class SuperTests(unittest.TestCase):
         second = next(ast_nodes[1].infer())
         self.assertIsInstance(second, bases.Instance)
 
+    def test_super_method_chaining_return_type(self) -> None:
+        """Test that super().method() correctly infers child type.
+
+        Regression test for: https://github.com/pylint-dev/astroid/issues/457
+        """
+        node = builder.extract_node(
+            """
+        class Parent:
+            def method(self) -> "Parent":
+                return self
+
+        class Child(Parent):
+            def method(self) -> "Child":
+                return super().method()
+
+            def extra(self) -> "Child":
+                return self
+
+        Child().method().extra() #@
+        """
+        )
+        inferred = next(node.infer())
+        # Should infer Child instance, not Parent
+        self.assertIsInstance(inferred, bases.Instance)
+        self.assertEqual(inferred._proxied.name, "Child")
+
+    def test_super_classmethod_chaining_return_type(self) -> None:
+        """Test that super().classmethod() correctly infers child type.
+
+        Regression test for: https://github.com/pylint-dev/astroid/issues/457
+        """
+        node = builder.extract_node(
+            """
+        class Parent:
+            @classmethod
+            def create(cls) -> "Parent":
+                return cls()
+
+        class Child(Parent):
+            @classmethod
+            def create(cls) -> "Child":
+                return super().create()
+
+            def extra(self) -> "Child":
+                return self
+
+        Child.create().extra() #@
+        """
+        )
+        inferred = next(node.infer())
+        # Should infer Child instance, not Parent
+        self.assertIsInstance(inferred, bases.Instance)
+        self.assertEqual(inferred._proxied.name, "Child")
+
     def test_super_invalid_types(self) -> None:
         node = builder.extract_node(
             """

